### PR TITLE
Add DNS provider for ArvanCloud

### LINKED ISSUE #1170: [Feature Request] Add support for ArvanCloud
ArvanCloud is a DNS and CDN provider.
you can find DNS API here: https://arvancloud.com/docs/api/cdn/4.0#tag/DNS

I'd add it if that's ok

### PR DESCRIPTION
**Pull request by sijad**
closes #1170

### CODE REVIEW

**Comment by sijad** [comment]
it's ready to be reviewed, I think CI is failing is unrelated to this changes.
as there's another PR for ArvanCloud, please let me know if I should make this PR undraft.

**Comment by ldez** [comment]
Hello, in order for a PR adding a DNS provider to be accepted, you have to:

- [ ] add a description to your PR
- [ ] be able to maintain this provider
- [ ] have a homogeneous design with the other providers
- [ ] add tests (units)
```shell
make test
```
- [ ] add tests ("live")
https://github.com/go-acme/lego/blob/c938de68f262c057542d44bcb67ef961cefcc60e/providers/dns/cloudflare/cloudflare_test.go#L125-L151

```shell
make test
```
- [ ] add the provider to the [readme.md](https://github.com/go-acme/lego#dns-providers)
- [ ] add a [provider descriptor](https://github.com/go-acme/lego/blob/master/providers/dns/cloudflare/cloudflare.toml)
- [ ] generate CLI help and documentation
```bash
make generate-dns
```
- [ ] be able to do: _(and put the output of this command to a comment in your PR)_
```bash
rm -rf .lego

./lego -m your@email.com --dns YOUR_PROVIDER_NAME -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```
Note the wildcard domain is important.
- [ ] pass the linter ([golangci-lint](https://github.com/golangci/golangci-lint#install) must be installed):
```shell
make checks
```
- [ ] do `go mod tidy`

**Comment by ldez** [comment]
@sijad and @FaRzad-845, could you do : (and put the output of this command to a comment in the PR)

```bash
// replace `your@email.com` by your email.
// replace `example.com` by your domain.

rm -rf .lego

./lego -m your@email.com --dns arvancloud -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```

Note the wildcard domain is important.

**Comment by sijad** [comment]
env ARVANCLOUD_API_KEY='Apikey xxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx' ./lego -a --dns arvancloud -d \*.thetangle.ir -d thetangle.ir --email hi@hashemian.me -s https://acme-staging-v02.api.letsencrypt.org/directory run
2020/05/28 17:43:45 No key found for account hi@hashemian.me. Generating a P384 key.
2020/05/28 17:43:45 Saved key to /home/jimc/go/src/github.com/go-acme/lego/cmd/lego/.lego/accounts/acme-staging-v02.api.letsencrypt.org/hi@hashemian.me/keys/hi@hashemian.me.key
2020/05/28 17:43:50 [INFO] acme: Registering account for hi@hashemian.me
!!!! HEADS UP !!!!

Your account credentials have been saved in your Let's Encrypt
configuration directory at "/home/jimc/go/src/github.com/go-acme/lego/cmd/lego/.lego/accounts".

You should make a secure backup of this folder now. This
configuration directory will also contain certificates and
private keys obtained from Let's Encrypt so making regular
backups of this folder is ideal.
2020/05/28 17:43:51 [INFO] [*.thetangle.ir, thetangle.ir] acme: Obtaining bundled SAN certificate
2020/05/28 17:43:53 [INFO] [*.thetangle.ir] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/59926102
2020/05/28 17:43:53 [INFO] [thetangle.ir] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/59926103
2020/05/28 17:43:53 [INFO] [*.thetangle.ir] acme: use dns-01 solver
2020/05/28 17:43:53 [INFO] [thetangle.ir] acme: Could not find solver for: tls-alpn-01
2020/05/28 17:43:53 [INFO] [thetangle.ir] acme: Could not find solver for: http-01
2020/05/28 17:43:53 [INFO] [thetangle.ir] acme: use dns-01 solver
2020/05/28 17:43:53 [INFO] [*.thetangle.ir] acme: Preparing to solve DNS-01
2020/05/28 17:44:11 [INFO] [thetangle.ir] acme: Preparing to solve DNS-01
2020/05/28 17:44:13 [INFO] [*.thetangle.ir] acme: Trying to solve DNS-01
2020/05/28 17:44:13 [INFO] [*.thetangle.ir] acme: Checking DNS record propagation using [8.8.8.8:53 8.8.4.4:53 192.168.43.202:53]
2020/05/28 17:44:13 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2020/05/28 17:44:16 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 17:44:20 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 17:44:24 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 17:44:27 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 17:44:30 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 17:44:34 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 17:44:37 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 17:44:41 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 17:44:44 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 17:44:48 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 17:44:51 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 17:44:54 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 17:44:59 [INFO] [*.thetangle.ir] The server validated our request
2020/05/28 17:44:59 [INFO] [thetangle.ir] acme: Trying to solve DNS-01
2020/05/28 17:44:59 [INFO] [thetangle.ir] acme: Checking DNS record propagation using [8.8.8.8:53 8.8.4.4:53 192.168.43.202:53]
2020/05/28 17:44:59 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2020/05/28 17:45:01 [INFO] [thetangle.ir] The server validated our request
2020/05/28 17:45:01 [INFO] [*.thetangle.ir] acme: Cleaning DNS-01 challenge
2020/05/28 17:45:04 [INFO] [thetangle.ir] acme: Cleaning DNS-01 challenge
2020/05/28 17:45:15 [INFO] [*.thetangle.ir, thetangle.ir] acme: Validations succeeded; requesting certificates
2020/05/28 17:45:17 [INFO] [*.thetangle.ir] Server responded with a certificate.

**Comment by ldez** [comment]
@sijad you have pushed force on my changes, I restored the branch.

Could you use the current branch and retry.

Please don't squash or push force the branch, thank you.

**Comment by sijad** [comment]
I get this error now:

```sh
2020/05/28 18:14:45 No key found for account hi@hashemian.me. Generating a P384 key.
2020/05/28 18:14:45 Saved key to /home/jimc/go/src/github.com/go-acme/lego/cmd/lego/.lego/accounts/acme-staging-v02.api.letsencrypt.org/hi@hashemian.me/keys/hi@hashemian.me.key
2020/05/28 18:14:51 [INFO] acme: Registering account for hi@hashemian.me
!!!! HEADS UP !!!!

Your account credentials have been saved in your Let's Encrypt
configuration directory at "/home/jimc/go/src/github.com/go-acme/lego/cmd/lego/.lego/accounts".

You should make a secure backup of this folder now. This
configuration directory will also contain certificates and
private keys obtained from Let's Encrypt so making regular
backups of this folder is ideal.
2020/05/28 18:14:52 [INFO] [*.thetangle.ir, thetangle.ir] acme: Obtaining bundled SAN certificate
2020/05/28 18:14:54 [INFO] [*.thetangle.ir] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/59933345
2020/05/28 18:14:54 [INFO] [thetangle.ir] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/59933346
2020/05/28 18:14:54 [INFO] [*.thetangle.ir] acme: use dns-01 solver
2020/05/28 18:14:54 [INFO] [thetangle.ir] acme: Could not find solver for: tls-alpn-01
2020/05/28 18:14:54 [INFO] [thetangle.ir] acme: Could not find solver for: http-01
2020/05/28 18:14:54 [INFO] [thetangle.ir] acme: use dns-01 solver
2020/05/28 18:14:54 [INFO] [*.thetangle.ir] acme: Preparing to solve DNS-01
panic: assignment to entry in nil map

goroutine 1 [running]:
github.com/go-acme/lego/v3/providers/dns/arvancloud.(*DNSProvider).Present(0xc0004dfd20, 0xc000046db0, 0xc, 0xc0005a80c0, 0x2b, 0xc000628000, 0x57, 0xc000046da8, 0x3)
	/home/jimc/go/src/github.com/go-acme/lego/providers/dns/arvancloud/arvancloud.go:132 +0x2de
github.com/go-acme/lego/v3/challenge/dns01.(*Challenge).PreSolve(0xc00036ab40, 0xc000046dd0, 0x7, 0x0, 0xed66af1d4, 0x0, 0xc000046da8, 0x3, 0xc000046db0, 0xc, ...)
	/home/jimc/go/src/github.com/go-acme/lego/challenge/dns01/dns_challenge.go:94 +0x2a1
github.com/go-acme/lego/v3/challenge/resolver.parallelSolve(0xc0004ae240, 0x2, 0x2, 0xc000439ad0)
	/home/jimc/go/src/github.com/go-acme/lego/challenge/resolver/prober.go:135 +0x11c
github.com/go-acme/lego/v3/challenge/resolver.(*Prober).Solve(0xc0000102d8, 0xc0001a0000, 0x2, 0x2, 0x1e, 0xc00018c080)
	/home/jimc/go/src/github.com/go-acme/lego/challenge/resolver/prober.go:84 +0x359
github.com/go-acme/lego/v3/certificate.(*Certifier).Obtain(0xc0004cb530, 0xc00036db80, 0x2, 0x2, 0x1, 0x0, 0x0, 0x0, 0x199, 0xc00059e080, ...)
	/home/jimc/go/src/github.com/go-acme/lego/certificate/certificates.go:114 +0x2da
github.com/go-acme/lego/v3/cmd.obtainCertificate(0xc000474160, 0xc0004dfd00, 0x164fc69, 0x145, 0xc000565768)
	/home/jimc/go/src/github.com/go-acme/lego/cmd/cmd_run.go:166 +0x2d8
github.com/go-acme/lego/v3/cmd.run(0xc000474160, 0x0, 0x2)
	/home/jimc/go/src/github.com/go-acme/lego/cmd/cmd_run.go:84 +0xe6
github.com/urfave/cli.HandleAction(0x12fab00, 0x167a6f8, 0xc000474160, 0xc000474160, 0x0)
	/home/jimc/go/pkg/mod/github.com/urfave/cli@v1.22.1/app.go:523 +0x11a
github.com/urfave/cli.Command.Run(0x15f2212, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x163e0f8, 0x3a, 0x0, ...)
	/home/jimc/go/pkg/mod/github.com/urfave/cli@v1.22.1/command.go:174 +0x56e
github.com/urfave/cli.(*App).Run(0xc00046a000, 0xc00011c680, 0xd, 0xd, 0x0, 0x0)
	/home/jimc/go/pkg/mod/github.com/urfave/cli@v1.22.1/app.go:276 +0x7ae
main.main()
```

**Comment by ldez** [comment]
I will fix that

**Comment by ldez** [comment]
Fixed, could you retry?

File: providers/dns/arvancloud/arvancloud.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+	// deletes record ID from map
+	d.recordIDsMu.Lock()
+	delete(d.recordIDs, token)
+	d.recordIDsMu.Unlock()
+
+	return nil
+}
+
+func getZone(fqdn string) (string, error) {
```
Comment (sijad) [inline providers/dns/arvancloud/arvancloud.go:171]: just asking, doesn't [effective go](https://golang.org/doc/effective_go.html#Getters) discourages this kind of naming?

**Comment by sijad** [comment]
now I get this error:

```
2020/05/28 18:23:27 [INFO] [*.thetangle.ir, thetangle.ir] acme: Obtaining bundled SAN certificate
2020/05/28 18:23:30 [INFO] [*.thetangle.ir] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/59934974
2020/05/28 18:23:30 [INFO] [thetangle.ir] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/59934975
2020/05/28 18:23:30 [INFO] [*.thetangle.ir] acme: use dns-01 solver
2020/05/28 18:23:30 [INFO] [thetangle.ir] acme: Could not find solver for: tls-alpn-01
2020/05/28 18:23:30 [INFO] [thetangle.ir] acme: Could not find solver for: http-01
2020/05/28 18:23:30 [INFO] [thetangle.ir] acme: use dns-01 solver
2020/05/28 18:23:30 [INFO] [*.thetangle.ir] acme: Preparing to solve DNS-01
2020/05/28 18:23:50 [INFO] [thetangle.ir] acme: Preparing to solve DNS-01
2020/05/28 18:23:51 [INFO] [*.thetangle.ir] acme: Trying to solve DNS-01
2020/05/28 18:23:51 [INFO] [*.thetangle.ir] acme: Checking DNS record propagation using [8.8.8.8:53 8.8.4.4:53 192.168.43.202:53]
2020/05/28 18:23:51 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2020/05/28 18:24:02 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:24:15 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:24:28 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:24:41 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:24:50 [INFO] [*.thetangle.ir] The server validated our request
2020/05/28 18:24:50 [INFO] [thetangle.ir] acme: Trying to solve DNS-01
2020/05/28 18:24:50 [INFO] [thetangle.ir] acme: Checking DNS record propagation using [8.8.8.8:53 8.8.4.4:53 192.168.43.202:53]
2020/05/28 18:24:50 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2020/05/28 18:24:58 [INFO] [thetangle.ir] The server validated our request
2020/05/28 18:24:58 [INFO] [*.thetangle.ir] acme: Cleaning DNS-01 challenge
2020/05/28 18:25:00 [WARN] [*.thetangle.ir] acme: cleaning up failed: arvancloud: failed to delate TXT record: id=4c40e758-6b94-4a34-aa77-0909232744ac: could not delete record 4c40e758-6b94-4a34-aa77-0909232744ac; Domain: thetangle.ir; Status: 500 Internal Server Error 
2020/05/28 18:25:00 [INFO] [thetangle.ir] acme: Cleaning DNS-01 challenge
2020/05/28 18:25:01 [WARN] [thetangle.ir] acme: cleaning up failed: arvancloud: failed to delate TXT record: id=7569d7e3-e547-4541-91e9-206fe7883ea2: could not delete record 7569d7e3-e547-4541-91e9-206fe7883ea2; Domain: thetangle.ir; Status: 500 Internal Server Error 
2020/05/28 18:25:01 [INFO] [*.thetangle.ir, thetangle.ir] acme: Validations succeeded; requesting certificates
2020/05/28 18:25:03 [INFO] [*.thetangle.ir] Server responded with a certificate.
```

please note `search` parameter can not contain `_`.

https://github.com/go-acme/lego/commit/9ca50e9963c620ace526e870fcee526334d7a613#diff-90d063d641bee15c29d584998b371f9fR38

BTW I'm sorry about your changes, I didn't noticed them

**Comment by sijad** [comment]
500 error can be fixed by this: https://github.com/go-acme/lego/commit/9ca50e9963c620ace526e870fcee526334d7a613#diff-90d063d641bee15c29d584998b371f9fR137-R139

File: providers/dns/arvancloud/arvancloud.go (hunk 1 above)
Comment (ldez) [reply providers/dns/arvancloud/arvancloud.go:171]: It's not a real getters (no struct, no method, no attribute related to an attribute)

**Comment by ldez** [comment]
the `GetTxtRecord` is not used, replace it by a map of IDs.

**Comment by ldez** [comment]
I improved the error log, could you retry?

File: providers/dns/arvancloud/internal/client.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+func (c *Client) do(method, endpoint string, body io.Reader) (*http.Response, error) {
+	req, err := http.NewRequest(method, endpoint, body)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("Content-Type", "application/json")
```
Comment (sijad) [inline providers/dns/arvancloud/internal/client.go:165]: this is cause the 500 error, `GET` should not have this header.

```
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
```

Comment (ldez) [reply providers/dns/arvancloud/internal/client.go:165]: There is no GET so the error is not related to that.

Comment (ldez) [reply providers/dns/arvancloud/internal/client.go:165]: I improved the error log, could you retry?

**Comment by sijad** [comment]
> the `GetTxtRecord` is not used, I replaced it by a map of IDs.
> 
> so the 500 cannot be fixed by changing that.

yes that was my mistake but challenges sometimes contains `_` and it cause an error too:

```json
{"status":false,"message":"The given data was invalid.","errors":{"search":["The search is invalid."]}}
```

 e.g. `https://napi.arvancloud.com/cdn/4.0/domains/thetangle.ir/dns-records?search=YhQB5ZCbhy87MPUUvd2INfU2ZPOLD4RUumZ6Kd7B_TQ`

**Comment by ldez** [comment]
I improved the error log, could you retry?

File: providers/dns/arvancloud/internal/client.go (hunk 2 above)
Comment (FaRzad-845) [reply providers/dns/arvancloud/internal/client.go:165]: I may be wrong, but what I saw was that if we want to delete a record and the given id is not valid, arvancloud return 500 page .

Comment (sijad) [reply providers/dns/arvancloud/internal/client.go:165]: here:

```
2020/05/28 18:54:18 No key found for account hi@hashemian.me. Generating a P384 key.
2020/05/28 18:54:18 Saved key to /home/jimc/go/src/github.com/go-acme/lego/cmd/lego/.lego/accounts/acme-staging-v02.api.letsencrypt.org/hi@hashemian.me/keys/hi@hashemian.me.key
2020/05/28 18:54:21 [INFO] acme: Registering account for hi@hashemian.me
!!!! HEADS UP !!!!

Your account credentials have been saved in your Let's Encrypt
configuration directory at "/home/jimc/go/src/github.com/go-acme/lego/cmd/lego/.lego/accounts".

You should make a secure backup of this folder now. This
configuration directory will also contain certificates and
private keys obtained from Let's Encrypt so making regular
backups of this folder is ideal.
2020/05/28 18:54:22 [INFO] [*.thetangle.ir, thetangle.ir] acme: Obtaining bundled SAN certificate
2020/05/28 18:54:25 [INFO] [*.thetangle.ir] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/59941157
2020/05/28 18:54:25 [INFO] [thetangle.ir] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/59941158
2020/05/28 18:54:25 [INFO] [*.thetangle.ir] acme: use dns-01 solver
2020/05/28 18:54:25 [INFO] [thetangle.ir] acme: Could not find solver for: tls-alpn-01
2020/05/28 18:54:25 [INFO] [thetangle.ir] acme: Could not find solver for: http-01
2020/05/28 18:54:25 [INFO] [thetangle.ir] acme: use dns-01 solver
2020/05/28 18:54:25 [INFO] [*.thetangle.ir] acme: Preparing to solve DNS-01
2020/05/28 18:54:38 [INFO] [thetangle.ir] acme: Preparing to solve DNS-01
2020/05/28 18:54:44 [INFO] [*.thetangle.ir] acme: Trying to solve DNS-01
2020/05/28 18:54:44 [INFO] [*.thetangle.ir] acme: Checking DNS record propagation using [8.8.8.8:53 8.8.4.4:53 192.168.43.202:53]
2020/05/28 18:54:44 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2020/05/28 18:54:46 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:54:50 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:54:54 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:54:58 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:55:03 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:55:07 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:55:11 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:55:14 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:55:17 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:55:24 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:55:29 [INFO] [*.thetangle.ir] The server validated our request
2020/05/28 18:55:29 [INFO] [thetangle.ir] acme: Trying to solve DNS-01
2020/05/28 18:55:29 [INFO] [thetangle.ir] acme: Checking DNS record propagation using [8.8.8.8:53 8.8.4.4:53 192.168.43.202:53]
2020/05/28 18:55:29 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2020/05/28 18:55:32 [INFO] [thetangle.ir] The server validated our request
2020/05/28 18:55:32 [INFO] [*.thetangle.ir] acme: Cleaning DNS-01 challenge
2020/05/28 18:55:33 [WARN] [*.thetangle.ir] acme: cleaning up failed: arvancloud: failed to delate TXT record: id=5ace4bfb-014c-4362-8f40-0e4e50569118: could not delete record 5ace4bfb-014c-4362-8f40-0e4e50569118; Domain: thetangle.ir; Status: 500 Internal Server Error; Body: {
    "status": false,
    "message": "Internal Server Error"
} 
2020/05/28 18:55:33 [INFO] [thetangle.ir] acme: Cleaning DNS-01 challenge
2020/05/28 18:55:34 [WARN] [thetangle.ir] acme: cleaning up failed: arvancloud: failed to delate TXT record: id=27e49a84-1df4-4bac-ac96-8d7f33cfcb64: could not delete record 27e49a84-1df4-4bac-ac96-8d7f33cfcb64; Domain: thetangle.ir; Status: 500 Internal Server Error; Body: {
    "status": false,
    "message": "Internal Server Error"
} 
2020/05/28 18:55:34 [INFO] [*.thetangle.ir, thetangle.ir] acme: Validations succeeded; requesting certificates
2020/05/28 18:55:36 [INFO] [*.thetangle.ir] Server responded with a certificate.
```

I'm pretty sure my suggestion fixes it

Comment (ldez) [reply providers/dns/arvancloud/internal/client.go:165]: There are no GET request so it's impossible.

But maybe you are speaking about DELETE, but I have not changed this part in my review/merge.

Comment (ldez) [reply providers/dns/arvancloud/internal/client.go:165]: > I may be wrong, but what I saw was that if we want to delete a record and the given id is not valid, arvancloud return 500 page .

I think that @FaRzad-845 is right.

Comment (FaRzad-845) [reply providers/dns/arvancloud/internal/client.go:165]: i add this line in my codes 

```if method != http.MethodDelete {
		req.Header.Set("Content-Type", "application/json")
	}```

Comment (sijad) [reply providers/dns/arvancloud/internal/client.go:165]: yes, I meant `DELETE`, that was my mistake, I've fixed it later (which accidentally removed your changes too)

Comment (ldez) [reply providers/dns/arvancloud/internal/client.go]: I fixed that, could you retry once again?

**Comment by FaRzad-845** [comment]
i think you should use != instead of == @ldez

**Comment by ldez** [comment]
@FaRzad-845  fixed

**Comment by ldez** [comment]
@sijad or @FaRzad-845  could you retry?

**Comment by sijad** [comment]
Thanks it seems fixed now!

I wish I've seen your changes and didn't waste so many time!

```sh
env ARVANCLOUD_API_KEY='Apikey xxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx' ./lego -a --dns arvancloud -d \*.thetangle.ir -d thetangle.ir --email hi@hashemian.me -s https://acme-staging-v02.api.letsencrypt.org/directory run
2020/05/28 19:21:06 [INFO] [*.thetangle.ir, thetangle.ir] acme: Obtaining bundled SAN certificate
2020/05/28 19:21:09 [INFO] [*.thetangle.ir] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/59946993
2020/05/28 19:21:09 [INFO] [thetangle.ir] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/59946994
2020/05/28 19:21:09 [INFO] [*.thetangle.ir] acme: use dns-01 solver
2020/05/28 19:21:09 [INFO] [thetangle.ir] acme: Could not find solver for: tls-alpn-01
2020/05/28 19:21:09 [INFO] [thetangle.ir] acme: Could not find solver for: http-01
2020/05/28 19:21:09 [INFO] [thetangle.ir] acme: use dns-01 solver
2020/05/28 19:21:09 [INFO] [*.thetangle.ir] acme: Preparing to solve DNS-01
2020/05/28 19:21:24 [INFO] [thetangle.ir] acme: Preparing to solve DNS-01
2020/05/28 19:21:26 [INFO] [*.thetangle.ir] acme: Trying to solve DNS-01
2020/05/28 19:21:26 [INFO] [*.thetangle.ir] acme: Checking DNS record propagation using [8.8.8.8:53 8.8.4.4:53 192.168.43.202:53]
2020/05/28 19:21:26 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2020/05/28 19:21:36 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 19:21:49 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 19:21:56 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 19:22:09 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 19:22:24 [INFO] [*.thetangle.ir] The server validated our request
2020/05/28 19:22:24 [INFO] [thetangle.ir] acme: Trying to solve DNS-01
2020/05/28 19:22:24 [INFO] [thetangle.ir] acme: Checking DNS record propagation using [8.8.8.8:53 8.8.4.4:53 192.168.43.202:53]
2020/05/28 19:22:24 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2020/05/28 19:22:28 [INFO] [thetangle.ir] The server validated our request
2020/05/28 19:22:28 [INFO] [*.thetangle.ir] acme: Cleaning DNS-01 challenge
2020/05/28 19:22:29 [INFO] [thetangle.ir] acme: Cleaning DNS-01 challenge
2020/05/28 19:22:31 [INFO] [*.thetangle.ir, thetangle.ir] acme: Validations succeeded; requesting certificates
2020/05/28 19:22:33 [INFO] [*.thetangle.ir] Server responded with a certificate.
```

**Pull request merged**